### PR TITLE
8344064: Remove doPrivileged calls from print/imageio/media classes in the java.desktop module

### PR DESCRIPTION
**Pull request by prrace**
Remove doPrivileged calls from javax.imageio, javax.print and some sound impl classes

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8344064](https://bugs.openjdk.org/browse/JDK-8344064): Remove doPrivileged calls from print/imageio/media classes in the java.desktop module (**Bug** - P4)

### Reviewers
 * [Alexander Zuev](https://openjdk.org/census#kizune) (@azuev-java - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22186/head:pull/22186` \
`$ git checkout pull/22186`

Update a local copy of the PR: \
`$ git checkout pull/22186` \
`$ git pull https://git.openjdk.org/jdk.git pull/22186/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22186`

View PR using the GUI difftool: \
`$ git pr show -t 22186`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22186.diff">https://git.openjdk.org/jdk/pull/22186.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22186#issuecomment-2484085017)
</details>

### CODE REVIEW

File: src/java.desktop/share/classes/javax/print/PrintServiceLookup.java [hunk 1]
```diff
@@ -344,32 +343,11 @@ private static ArrayList<PrintServiceLookup> getAllLookupServices() {
...
-                                }
-                            }
-                            return null;
-                        }
-                });
-            } catch (java.security.PrivilegedActionException e) {
+            Iterator<PrintServiceLookup> iterator = ServiceLoader.load(PrintServiceLookup.class).iterator();
+            ArrayList<PrintServiceLookup> los = getListOfLookupServices();
+            while (iterator.hasNext()) {
+                los.add(iterator.next());
```
Comment (TejeshR13) [inline src/java.desktop/share/classes/javax/print/PrintServiceLookup.java:349]: Any reason for removing the try-catch block which is handling `ServiceConfigurationError` exception?

File: src/java.desktop/share/classes/javax/imageio/metadata/IIOMetadata.java [hunk 2]
```diff
@@ -398,12 +396,9 @@ public IIOMetadataFormat getMetadataFormat(String formatName) {
             throw new IllegalArgumentException("Unsupported format name");
         }
         try {
-            final String className = formatClassName;
             // Try to load from the module of the IIOMetadata implementation
             // for this plugin since the IIOMetadataImpl is part of the plugin
-            PrivilegedAction<Class<?>> pa = () -> { return getMetadataFormatClass(className); };
-            @SuppressWarnings("removal")
-            Class<?> cls = AccessController.doPrivileged(pa);
+            Class<?> cls =  getMetadataFormatClass(formatClassName);
```
Comment (turbanoff) [inline src/java.desktop/share/classes/javax/imageio/metadata/IIOMetadata.java:401]:
```suggestion
            Class<?> cls = getMetadataFormatClass(formatClassName);
```

Comment (prrace) [reply src/java.desktop/share/classes/javax/imageio/metadata/IIOMetadata.java]: ok. But do you have any comments on the *functional* aspect of this PR ?

File: src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java [hunk 3]
```diff
@@ -634,15 +632,19 @@ public boolean remapInstrument(Instrument from, Instrument to) {
         }
     }
 
+   static interface RunnableAction<T> {
```
Comment (turbanoff) [inline src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java:635]: Can we use `.java.util.functionSupplier`  instead?

File: src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java [hunk 4]
```diff
@@ -798,7 +798,7 @@ public InputStream run() {
                             } catch (final FileNotFoundException ignored) {
                             }
                             return null;
-                        });
+                        }).run();
```
Comment (turbanoff) [inline src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java:801]: Let's unwrap this `RunnableAction` and put code directly in the `getDefaultSoundbank`

File: src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/Metacity.java [hunk 5]
```diff
@@ -506,21 +504,12 @@ void paintFrameBorder(SynthContext context, Graphics g, int x0, int y0, int widt
...
 
-        @SuppressWarnings("removal")
-        public Object doPrivileged(int type, Object arg) {
-            this.type = type;
-            this.arg = arg;
-            return AccessController.doPrivileged(this);
-        }
-
-        public Object run() {
+        public Object getThemeItem(int type, Object arg) {
```
[the remark this answers is not in the document]
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/java/swing/plaf/gtk/Metacity.java:512]: No. Refactoring is explicitly a non-goal in these PRs. And minimising required refactoring is a goal.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java (hunk 4 above)
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java:801]: No. Refactoring is explicitly a non-goal in these PRs. And minimising required refactoring is a goal.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java (hunk 3 above)
Comment (prrace) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java:635]: I could, but I chose to do this instead.

File: src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java (hunk 4 above)
Comment (turbanoff) [reply src/java.desktop/share/classes/com/sun/media/sound/SoftSynthesizer.java:801]: Will you file a separate ticket to cleanup this?
Current code looks confusing. Creating lambda just to execute it doesn't make much sense.

File: src/java.desktop/share/classes/com/sun/imageio/plugins/bmp/BMPImageReader.java [hunk 6]
```diff
@@ -2039,17 +2037,9 @@ public void readAborted(ImageReader src) {}
...
         if (isLinkedProfileAllowed == null) {
-            PrivilegedAction<Boolean> a = new PrivilegedAction<Boolean>() {
-                @Override
-                public Boolean run() {
-                    return Boolean.
-                        getBoolean("sun.imageio.bmp.enableLinkedProfiles");
-                }
-            };
-            isLinkedProfileAllowed = AccessController.doPrivileged(a);
+            isLinkedProfileAllowed = Boolean.getBoolean("sun.imageio.bmp.enableLinkedProfiles");
```
Comment (azuev-java) [inline src/java.desktop/share/classes/com/sun/imageio/plugins/bmp/BMPImageReader.java:2042]: Do we even need that static variable? Why not just
`return Boolean.getBoolean("sun.imageio.bmp.enableLinkedProfiles");`
Performance? Does it deserve an internal caching? Just curious.

**Review by azuev-java — APPROVED** [review]
Other than a minor clarification in BMPImageReader that does not affect functionality changes look good.

File: src/java.desktop/share/classes/javax/print/StreamPrintServiceFactory.java [hunk 7]
```diff
@@ -183,43 +183,23 @@ private static ArrayList<StreamPrintServiceFactory> initListOfFactories() {
...
-                        }
-                });
-            } catch (java.security.PrivilegedActionException e) {
-            }
-            return listOfFactories;
+              Iterator<StreamPrintServiceFactory> iterator =
+                  ServiceLoader.load(StreamPrintServiceFactory.class).iterator();
+              ArrayList<StreamPrintServiceFactory> lof = getListOfFactories();
+              while (iterator.hasNext()) {
+                  lof.add(iterator.next());
```
Comment (TejeshR13) [inline src/java.desktop/share/classes/javax/print/StreamPrintServiceFactory.java:200]: `lof` is set with data here but never used. Was it used only to check for `ServiceConfigurationError` exception? or am I missing something?

**Pull request closed without merging**